### DISCUSSION TITLE
MQTT topic structure

### DISCUSSION THREAD

**Discussion by miccgn** [pixelit-project/PixelIt#137] [Ideas]
Hi,
at the moment I am playing around with MQTT.
The current structure offers
- a topic for the lux sensor with a JSON object containing the lux value
- a topic for other sensors with a JSON object containing up to four values
- a topic with general information about the PixelIt device

I'd love to implement more topics with a structure that's more easy to parse. There are two options I could think of:
### A) Separate topics only
keep the current structure and add separate topics for:
- lux sensor
- temperature sensor
- pressure sensor
- humidity sensor
- gas sensor

Matrix info will be available as JSON object only.

I wouldn't suggest any changes to the incoming message structure:
- setScreen will require a JSON object in any case
- getLux etc. are more or less useless, as MQTT messages will be posted every few seconds anyway

### B) [Homie convention](https://homieiot.github.io/)-like
keep the current structure and add separate Homie-like topics for:
- node "homie/<hostname>/sensors"
- homie/\<hostname>/sensors/luxsensor
- homie/\<hostname>/sensors/temperature
- homie/\<hostname>/sensors/pressure
- homie/\<hostname>/sensors/humidity
- homie/\<hostname>/sensors/gas
(plus all required $attributes as per Homie convention).
==> Matrix info will be available as JSON object only.

For the incoming messages, I would keep the current structure plus add a Homie-style topic for setScreen only (requiring a JSON object)

What do you think?

**Comment by miccgn** [comment]
The more I dive into the Homie convention, I begin to love this one :)

Refined proposal:
- create a WebUI option: "Use Homie topics", which disables the Topic setting and sending of the currently known topics.
- add new readable topics with information:
pixelit/\<HOSTNAME>/sensor/brightness
pixelit/\<HOSTNAME>/sensor/temperature (when TempSensor is present)
pixelit/\<HOSTNAME>/sensor/humidity (when TempSensor is present)
pixelit/\<HOSTNAME>/sensor/pressure (when BME280 or BME680 is present)
pixelit/\<HOSTNAME>/sensor/gas (when BME680 is present)
pixelit/\<HOSTNAME>/buttons/button1
pixelit/\<HOSTNAME>/buttons/button2
pixelit/\<HOSTNAME>/buttons/button3
pixelit/\<HOSTNAME>/core/version
pixelit/\<HOSTNAME>/core/hostname
pixelit/\<HOSTNAME>/core/ip
pixelit/\<HOSTNAME>/core/note
pixelit/\<HOSTNAME>/core/config (JSON object)
pixelit/\<HOSTNAME>/screen/sleepmode
pixelit/\<HOSTNAME>/screen/brightness
pixelit/\<HOSTNAME>/$state (as per Homie convention this will be the Last Will Topic)

- add new settable topics:
Add following settable topics:
pixelit/\<HOSTNAME>/core/config (expects JSON)
pixelit/\<HOSTNAME>/screen/screendef (expects JSON)

What do you think?

**Comment by o0shojo0o** [comment]
I don't use mqtt myself, so i can't really have an opinion.  
However, if it is so better for you to handle, that can be included with pleasure!

**Comment by miccgn** [comment]
Hi!

There's one thing that didn't feel good with Homie: I would have to alter the topic structure currently being published.

Seems a lot of people use HomeAssistant, and many devices seem to support the discovery structure HomeAssistant suggests. Given that there's support for HomeAssistant discovery in OpenHAB and ioBroker (and probably in some more systems), and given that the HomeAssistant discovery would allow to keep the current topic structure, why not use that one?

Therefore I decided to add discovery information according to the [HomeAssistant MQTT discovery documentation](https://www.home-assistant.io/docs/mqtt/discovery/). 

As I am not using HomeAssistant, I performed tests with OpenHAB. The mqtt add-on of OpenHAB supports HomeAssistant discovery. Everything seems to work fine by now.

Please feel free to test!

**Reply by foorschtbar** [reply]
I don't use HassIO (I can do everything with Node-RED, so why do I need HassIO?), but your changes look very good. Thanks for your contribution.
